test/unit/math/prim/prob/hmm_marginal_test.cpp

```diff
@@ -0,0 +1,243 @@
+    rho(i) = rho_unconstrained[i];
+
+  return stan::math::hmm_marginal_lpdf(log_omegas, Gamma, rho);
+}
+
+/**
+ * In the proposed example, the latent sate x determines
+ * the observational distribution:
+ *  (i) normal(mu, sigma)
+ *  (ii) normal(-mu, sigma)
+ */
+double state_lpdf(double y, double abs_mu, double sigma, int state) {
+  int x = (-2 * state + 1);
+  double chi = (y - x * abs_mu) / sigma;
+  return -0.5 * chi * chi - 0.5 * std::log(6.283185307179586) - std::log(sigma);
+}
+
+class hmm_marginal_lpdf_test : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    n_states = 2;
+    p1_init = 0.65;
+    gamma1 = 0.7;
+    gamma2 = 0.45;
+    n_transitions = 10;
+    abs_mu = 1;
+    sigma = 1;
+
+    Eigen::VectorXd rho_(n_states);
+    rho_ << p1_init, 1 - p1_init;
+    rho = rho_;
+
+    Eigen::MatrixXd Gamma_(n_states, n_states);
+    Gamma_ << gamma1, 1 - gamma1, gamma2, 1 - gamma2;
+    Gamma = Gamma_;
+
+    Eigen::VectorXd obs_data_(n_transitions + 1);
+    obs_data_ << -0.3315914, -0.1655340, -0.7984021, 0.2364608, -0.4489722,
+        2.1831438, -1.4778675, 0.8717423, -1.0370874, 0.1370296, 1.9786208;
+    obs_data = obs_data_;
+
+    Eigen::MatrixXd log_omegas_(n_states, n_transitions + 1);
+    for (int n = 0; n < n_transitions + 1; n++) {
+      log_omegas_.col(n)[0] = state_lpdf(obs_data[n], abs_mu, sigma, 0);
+      log_omegas_.col(n)[1] = state_lpdf(obs_data[n], abs_mu, sigma, 1);
+    }
+    log_omegas = log_omegas_;
+    log_omegas_zero = log_omegas.block(0, 0, n_states, 1);
+
+    std::vector<double> rho_unconstrained_(n_states - 1);
+    for (int i = 0; i < rho.size() - 1; i++)
+      rho_unconstrained_[i] = rho(i);
+    rho_unconstrained = rho_unconstrained_;
+
+    Gamma_unconstrained = Gamma.block(0, 0, n_states, n_states - 1);
+  }
+
+  int n_states, n_transitions;
+  double abs_mu, sigma, p1_init, gamma1, gamma2;
+
+  Eigen::VectorXd rho;
+  Eigen::MatrixXd Gamma;
+  Eigen::VectorXd obs_data;
+  Eigen::MatrixXd log_omegas;
+  Eigen::MatrixXd log_omegas_zero;
+
+  // Construct "unconstrained" versions of rho and Gamma, without
+  // the final element which can be determnied using the fact
+  // the columns sum to 1. This allows us to do finite diff tests,
+  // without violating the simplex constraint of rho and Gamma.
+  std::vector<double> rho_unconstrained;
+  Eigen::MatrixXd Gamma_unconstrained;
+  stan::test::ad_tolerances tols;
+};
+
+TEST_F(hmm_marginal_lpdf_test, ten_transitions) {
+  using stan::math::hmm_marginal_lpdf;
+
+  EXPECT_FLOAT_EQ(-18.37417, hmm_marginal_lpdf(log_omegas, Gamma, rho));
+
+  // Differentiation tests
+  auto hmm_functor = [](const auto& log_omegas, const auto& Gamma_unconstrained,
+                        const auto& rho_unconstrained) {
+    return hmm_marginal_test_wrapper(log_omegas, Gamma_unconstrained,
+                                     rho_unconstrained);
+  };
+
+  stan::test::expect_ad(tols, hmm_functor, log_omegas, Gamma_unconstrained,
```

Comment:
[comment]
Cool!

Reply:
Cool indeed. 

But it's a pain to debug, because everything gets tested at once, and you get a mountain of error messages when something is wrong. It's also hard to pinpoint which test causes the error.

Reply:
I meant "cool it's passing the tests"!  I agree that the testing framework's hard to debug when things go wrong.  I think having a test_ad_var, test_ad_fvar_double, and so forth would help in order to run one test at a time.